### DISCUSSION TITLE
Doing lower() on get_by_email avoids index - SQLAlchemy

### DISCUSSION THREAD

**Discussion by AgustinFi** [fastapi-users/fastapi-users#1332] [Ideas]
Hi All,

I'd like to kick off a discussion about how get_by_email is querying the db (postgres in my case), specifically the part where lower() is applied to the email column. [Code Snippet](https://github.com/fastapi-users/fastapi-users-db-sqlalchemy/blob/f0078b049ed00d0399707d0c8fc35b6a4eb84751/fastapi_users_db_sqlalchemy/__init__.py#L125)  

```  
async def get_by_email(self, email: str) -> Optional[UP]:
        statement = select(self.user_table).where(
            func.lower(self.user_table.email) == func.lower(email)
        )
        return await self._get_user(statement)

```

- The index on that table is email, and not a function (lower(email)), so the lookup might be much less efficient.
- Emails, generally speaking, are case insensitive.

As long as the email gets stored in lower case, a possible fix would be:

```  
async def get_by_email(self, email: str) -> Optional[UP]:
        statement = select(self.user_table).where(
            self.user_table.email == func.lower(email)
        )
        return await self._get_user(statement)

```

Opinions?

**Comment by frankie567** [comment]
We store the email with the provided case *on purpose*. While generally, mails are case insensitive, things are not always that simple with emails: https://stackoverflow.com/questions/9807909/are-email-addresses-case-sensitive

So the choice we made here is to store the email as if (in case it actually matters for the recipient server), but have a lowercase lookup to avoid duplicates.

Regarding performance, I'm not convinced it's a big issue for most applications. If it is, PostgreSQL supports functional indices, which can easily be created if needed: https://www.postgresql.org/docs/current/indexes-expressional.html

**Comment by AgustinFi** [comment]
I was bringing this up in case this was not by design.

That being said, i would suggest creating a functional index just for the sake of clean code. 

Just my 2 cents.
